### ISSUE TITLE
Device admin permission to prevent uninstallation

### ISSUE THREAD

**Issue by fosshelper** [Alikaraki95/Getoffyourphone#5]
Thanks for this app. Please consider making the app require the device admin permission to prevent uninstallation when a block is going on. This would make such an app really effective.

**Comment by Alikaraki95** [comment]
Hey,

Thanks for the feedback & feature request.

Do you think that'll be accepted by everyon who's using it ? If i can make it selectable ( so people can enable/disable it whenever they want ) then yes, i will add it.

Please note: i develope this app by myself and i do it in my free time, updates/ new features might take some time.

Thanks again & regards,
Ali.

**Comment by fosshelper** [comment]
Please take your time. You are already doing us a service by providing an app of value in _your own_ time.

**Comment by Alikaraki95** [comment]
@fosshelper I've been doing a lot of research. Unfortunatly, you can't automatically activate/deactive Device admin permissions to disable uninstallation. So, i can't enable this whenver i want and disable it whenever i want ( like you suggested, prevent unisntallation only when the lock is running. ) 

Do you have any other requests :) ?

**Comment by fosshelper** [comment]
I mean, your app has to ask for permission to become a device administrator (much like other permissions). I don't think you can activate device admin permission whenever (you might be able to deactivate it whenever but I am not sure). In particular, one can keep a toggle box in Settings saying "Prevent uninstallation? Activate Device Admin..." and then your app could ask for device admin permission, and if granted, then your device can't be uninstalled without first disabling it as a device admin first.
For more info see [here](https://stackoverflow.com/questions/28794894/prevent-app-from-uninstall-in-android). In particular, a popular app called Self Focused (not open source) has this ability, you can have a look at it for reference.

**Comment by iexos** [comment]
Well, its possible to stop the lock effect by Force Stopping the app. I dont think there is any way around that, you will always be able to disable the lock if you really want to. I think of this app more as an impulse-stopper than a rootkit that is locking me out ;)

**Comment by Alikaraki95** [comment]
@iexos true, to some instant. I might be able to resume the timer after a force stop (doubt it) but it's a complex service, and i really need to research it before starting editing it :)

**Comment by Alikaraki95** [comment]
@fosshelper By settings app you mean for GOYP only or general phone settings? If so, I really doubt I'll be adding a feature that blocks settings app for everyone, ever if it's something that users opt in!

**Comment by fosshelper** [comment]
@Alikaraki95 The settings app (I mean general phone settings) is like any other app -- if you design your app to be able to block any app, the same applies to the settings app as well. Of course, users have the choice of blocking the settings app or not. 

Anyway, that's the only way this is going to be an effective block, since as @iexos pointed out, one can easily bypass the blocks by force stopping your app.

**Comment by Alikaraki95** [comment]
@fosshelper True, but i don't think it's a good aproach to Block Settings app - after all, if the app gets bugged or haywire for some reason, users will be blocked up and will need a phone restart, only because of GOYP. Get what i mean ?

I'll try a beta version of Admin Permissions and send it to you later, as i can't develop at the moment duo to personal reasons ( Life :( )

**Comment by fosshelper** [comment]
@Alikaraki95 I understand your concerns. It indeed does require a trust on the working of your app. But for better or for worse, that's how all self regulation software works. You give it higher privileges to be able to regulate you, but you need to trust that it won't seriously malfunction.

For advanced users, there are of course other ways to take care of such a situation. Once one has adb set up on one's computer and phone, one can use it to uninstall apps from the adb shell.

**Comment by Alikaraki95** [comment]
@fosshelper my only concern is the bug where the lock is broken/acts crazy and adds several hours when some user starts the lockdown BEFORE midnight 00:00, and this is a serious bug that i didn't find a solution for yet. It's related to the format + Time service, i know where the issue is but can't figure out an easy fix at the moment. When that is fixed, Admin Permissions will be officially released!
